crates/ide_completion/src/completions/keyword.rs

```diff
@@ -20,6 +20,10 @@ pub(crate) fn complete_expr_keyword(acc: &mut Completions, ctx: &CompletionConte
         cov_mark::hit!(no_keyword_completion_in_attr_of_expr);
         return;
     }
+    if ctx.is_non_trivial_path() {
+        cov_mark::hit!(no_keyword_completion_in_non_trivial_path);
+        return;
+    }
```

Comment:
I feel like it would be much cleaner to actually just check more explicitly where each keyword is applicable, instead of excluding a bunch of random situations here, but that's a refactoring for another day :sweat_smile: 

(I.e. something like "somewhere where an item could be defined" for stuff like `trait`)

Reply:
The completions already do this type of check in a lot of places, its also done here already. As you can see the incorrect completions were all item related, its just that the path prefix doesnt invalidate them due to how the checks work(basically the path that is being completed is looked at as a whole to figure out what kind of thing we are completing), so this check is technically correct for keyword completions here.

Reply:
I think there are some notes somewhere about how is this done in kotlin -- they insert the keyword, run the parser, check if there are errors, and have a pile of special-cases on top.

Reply:
@Veykril I know we're already doing this for a lot of checks, so we should take care to keep doing it that way ;) It seems to me that "at the beginning of what could be a path" and "in the middle of a qualified path" are fundamentally different locations to be completing for, and explicitly having to exclude "in the middle of a path" at the top of the function looks like a code smell.

Reply:
This is only the case for keyword completions though, for qualified and unqualified path completions for example the location is the same, item position.

So to me filtering in the keyword completions specifically seems correct.